bundles/org.openhab.binding.caddx/src/main/java/org/openhab/binding/caddx/internal/CaddxCommunicator.java

```diff
@@ -0,0 +1,521 @@
+        return serialPortName;
+    }
+
+    public int getBaudRate() {
+        return baudRate;
+    }
+
+    public void addListener(CaddxPanelListener listener) {
+        if (!listenerQueue.contains(listener)) {
+            listenerQueue.add(listener);
+        }
+    }
+
+    /**
+     * Send message to panel. Asynchronous, i.e. returns immediately.
+     * Messages are sent only when panel is ready (i.e. sent an
+     * acknowledgment to last message), but no checks are implemented that
+     * the message was correctly received and executed.
+     *
+     * @param msg Data to be sent to panel. First byte is message type.
+     *            Fletcher sum is computed and appended by transmit.
+     */
+    public void transmit(CaddxMessage msg) { // byte... msg) {
+        messages.add(msg);
+    }
+
+    /**
+     * Adds this message before any others in the queue.
+     * Used by receiver to send ACKs.
+     *
+     * @param msg The message
+     */
+    public void transmitFirst(CaddxMessage msg) {
+        messages.addFirst(msg);
+    }
+
+    public void stop() {
+        logger.trace("CaddxCommunicator stopping");
+
+        // kick thread out of waiting for FIFO
+        thread.interrupt();
+
+        // Close the streams first to unblock blocked reads and writes
+        try {
+            in.close();
+        } catch (IOException e) {
+        }
+        try {
+            out.close();
+        } catch (IOException e) {
+        }
+
+        // Wait until communication thread exits
+        try {
+            thread.join();
+        } catch (InterruptedException e) {
+        }
+
+        // Also close the serial port
+        serialPort.removeEventListener();
+        serialPort.close();
+    }
+
+    @SuppressWarnings("null")
+    @Override
+    public void run() {
+        int @Nullable [] expectedMessageNumbers = null;
+
+        @Nullable
+        CaddxMessage outgoingMessage = null;
+        boolean skipTransmit = true;
+
+        try {
+            // loop until the thread is interrupted, sending out messages
+            while (!Thread.currentThread().isInterrupted()) {
+                // Initialize the state
+                outgoingMessage = null;
+                expectedMessageNumbers = null;
+
+                if (!skipTransmit) {
+                    // send next outgoing message if we have one
+                    outgoingMessage = messages.poll();
+                    if (outgoingMessage != null) {
+                        logger.trace("CaddxCommunicator.run() Outgoing message: {}", outgoingMessage.getMessageType());
+                    }
+
+                    // Send the message
+                    if (outgoingMessage != null) {
+                        byte msg[] = outgoingMessage.getMessageFrameBytes(protocol);
+                        out.write(msg);
+                        out.flush();
+                    }
+
+                    // Log message
+                    if (outgoingMessage != null) {
```

Comment:
Instead of constantly checking this condition why not put them all in the same if statement block?

Reply:
Done